caikit/runtime/server_base.py

```diff
@@ -137,6 +137,43 @@ def interrupt(self, signal_, _stack_frame):
         )
         self.stop()
 
+    def _intercept_interrupt_signal(self) -> None:
+        """Intercept signal handlers to allow the server to stop on interrupt.
+        Calling this on a non-main thread has no effect.
+        This does not override any existing non-default signal handlers,
+        it will call them all in the reverse order they are registered.
+        """
+        self._add_signal_handler(signal.SIGINT, self.interrupt)
+        self._add_signal_handler(signal.SIGTERM, self.interrupt)
+
+    @staticmethod
+    def _add_signal_handler(sig, handler):
+        def nested_interrupt_builder(*handlers):
+            """Build and return an interrupt handler that calls all of *handlers"""
+
+            log.debug("Building interrupt handler: %s", handlers)
+
+            def interrupt(signal_, _stack_frame):
+                for handler in handlers:
+                    # Only call the handler if it is a callable fn that is _not_ a default handler
+                    log.debug("Running interrupt handler: %s", handler)
```

Comment:
Do we have to worry about writing to `stdout` in a signal handler? I know I've seen issues with writing to closed pipes before when logging in `__del__` functions

Reply:
uhhh, well we were logging before so I kept logging here? But that's a good question.

FWIW the uvicorn handler definitely spits out some info messages about shutting down 🤷 

Think this is relevant? https://docs.python.org/3/library/logging.html#thread-safety

> If you are implementing asynchronous signal handlers using the [signal](https://docs.python.org/3/library/signal.html#module-signal) module, you may not be able to use logging from within such handlers. This is because lock implementations in the [threading](https://docs.python.org/3/library/threading.html#module-threading) module are not always re-entrant, and so cannot be invoked from such signal handlers.

I don't exactly know if by "asynchronous" here they mean `async def ...` or just "asynchronous" in general 🤔